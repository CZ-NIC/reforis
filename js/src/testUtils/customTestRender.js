/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

/* eslint import/export: "off" */

import React from 'react';
import {UIDReset} from 'react-uid';
import {StaticRouter} from 'react-router';
import {render} from '@testing-library/react'

function Wrapper({children}) {
    return <StaticRouter>
        <UIDReset>
            {children}
        </UIDReset>
    </StaticRouter>
}

const customTestRender = (ui, options) => render(ui, {wrapper: Wrapper, ...options});

// re-export everything
export * from '@testing-library/react'

// override render method
export {customTestRender as render}
