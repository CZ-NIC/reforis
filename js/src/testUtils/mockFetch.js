/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

export default function mockFetch(data) {
    return jest.fn(() =>
        new Promise((resolve, reject) => {
            resolve({
                ok: true,
                json: () => {
                    return data ? data : {};
                },
            });
        })
    );
}
