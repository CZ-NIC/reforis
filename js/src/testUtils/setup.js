/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import mockAxios from "jest-mock-axios";
import moment from "moment-timezone";

// Setup axios cleanup
global.afterEach(() => {
    mockAxios.reset();
});

jest.doMock('moment', () => {
    moment.tz.setDefault('UTC');
    return moment;
});

jest.doMock("utils/vfs_fonts", () => ({}));

window.WebSocket = jest.fn();
