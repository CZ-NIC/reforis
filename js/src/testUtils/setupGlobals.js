/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

// Fake babel (gettext) used for docs
global._ = str => str;
global.babel = {format: (str) => str};
global.ForisTranslations = {};
