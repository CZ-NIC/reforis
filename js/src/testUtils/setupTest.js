/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import mockAxios from 'jest-mock-axios';
import moment from 'moment-timezone';

// Setup axios cleanup
global.afterEach(() => {
    mockAxios.reset();
});

// Mock babel (gettext)
window._ = str => str;
window.babel = {format: (str) => str};
window.ForisTranslations = {};

// Mock scrollIntoView
window.HTMLElement.prototype.scrollIntoView = () => {
};

jest.doMock('moment', () => {
    moment.tz.setDefault('UTC');
    return moment;
});

Date.now = jest.fn(() => new Date(Date.UTC(2019, 1, 1, 12, 13, 14)).valueOf());

jest.doMock('utils/vfs_fonts', () => {
    return {};
});
