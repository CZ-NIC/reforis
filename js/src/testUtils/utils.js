/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */


export function createPortalContainer(id) {
    let portalRoot = document.getElementById(id);
        if (!portalRoot) {
            portalRoot = document.createElement('div');
            portalRoot.setAttribute('id', id);
            document.body.appendChild(portalRoot)
        }
}
