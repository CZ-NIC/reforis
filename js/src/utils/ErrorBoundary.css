.error-boundary-description {
    display: block;
    background-color: #f2f2f2;
}
