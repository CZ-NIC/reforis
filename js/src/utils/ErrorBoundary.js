/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import PropTypes from "prop-types";

import { Portal } from "foris";

import "./ErrorBoundary.css";

export default class ErrorBoundary extends React.Component {
    constructor(props) {
        super(props);
        this.state = { error: null };
    }

    static getDerivedStateFromError(error) {
        return { error };
    }

    componentDidCatch(error, errorInfo) {
        // eslint-disable-next-line no-console
        console.error(
            `An error ocurred in reForis: "${error}", see details below.`,
            errorInfo.componentStack
        );
    }

    render() {
        const { error } = this.state;
        if (error) {
            document.title = "An Error Occurred - reForis";
            return (
                <Portal containerId="content-container">
                    <h1>{_("An Error Occurred")}</h1>
                    <code className="error-boundary-description p-3 my-3">
                        {error.toString()}
                    </code>
                    <p>
                        {_(
                            "More detailed information is available in the console of your web browser - on most browsers accessible after pressing Ctrl+Shift+J or F12."
                        )}
                    </p>
                    <p
                        dangerouslySetInnerHTML={{
                            __html: _(
                                'Please report this error to our support team via e-mail: <a href="mailto:tech.support@turris.cz">tech.support@turris.cz</a>.'
                            ),
                        }}
                    />
                </Portal>
            );
        }

        // eslint-disable-next-line react/destructuring-assignment
        return this.props.children;
    }
}

ErrorBoundary.propTypes = {
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node,
    ]),
};
