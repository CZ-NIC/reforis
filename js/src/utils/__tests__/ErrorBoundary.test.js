/*
 * Copyright (C) 2020 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";

import { useAPIGet } from "foris";
import { render } from "foris/testUtils/customTestRender";

import ErrorBoundary from  "../ErrorBoundary";

function mockBrokenGET() {
    throw new Error("This API request is broken!");
}

jest.mock("foris/api/utils", () => ({
    ...jest.requireActual('foris/api/utils'),
    API_METHODS: {
        GET: mockBrokenGET,
    },
}));

describe("<ErrorBoundary />", () => {
    it("should render children components when there's no error", () => {
        const { container } = render(
            <ErrorBoundary>
                <p>This is the content of the component.</p>
            </ErrorBoundary>
        );
        expect(container).toMatchSnapshot();
    });

    it("should display error message", () => {
        const BrokenComponent = () => {
            throw new Error("This component is broken!");
        };

        const content = document.createElement("div");
        content.setAttribute("id", "content-container");

        const originalError = console.error;
        console.error = jest.fn();
        const { container } = render(
            <>
                <ErrorBoundary>
                    <BrokenComponent />
                </ErrorBoundary>
            </>,
            { container: document.body.appendChild(content) }
        );
        expect(console.error).toBeCalled();
        console.error = originalError;

        expect(container).toMatchSnapshot();
    });

    it("should handle unexpected API error", () => {
        const BrokenAPIComponent = () => {
            const [, getExample] = useAPIGet("");
            useEffect(() => {
                getExample();
            }, [getExample]);
            return <p>No errors so far!</p>;
        };

        const content = document.createElement("div");
        content.setAttribute("id", "content-container");

        const originalError = console.error;
        console.error = jest.fn();
        const { container } = render(
            <>
                <ErrorBoundary>
                    <BrokenAPIComponent />
                </ErrorBoundary>
            </>,
            { container: document.body.appendChild(content) }
        );
        expect(console.error).toBeCalled();
        console.error = originalError;

        expect(container).toMatchSnapshot();
    });
});
