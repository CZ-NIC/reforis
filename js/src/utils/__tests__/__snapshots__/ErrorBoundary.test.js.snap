// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ErrorBoundary /> should display error message 1`] = `
<div
  id="content-container"
>
  <h1>
    An Error Occurred
  </h1>
  <code
    class="error-boundary-description p-3 my-3"
  >
    Error: This component is broken!
  </code>
  <p>
    More detailed information is available in the console of your web browser - on most browsers accessible after pressing Ctrl+Shift+J or F12.
  </p>
  <p>
    Please report this error to our support team via e-mail: 
    <a
      href="mailto:tech.support@turris.cz"
    >
      tech.support@turris.cz
    </a>
    .
  </p>
</div>
`;

exports[`<ErrorBoundary /> should handle unexpected API error 1`] = `
<div
  id="content-container"
>
  <h1>
    An Error Occurred
  </h1>
  <code
    class="error-boundary-description p-3 my-3"
  >
    Error: This API request is broken!
  </code>
  <p>
    More detailed information is available in the console of your web browser - on most browsers accessible after pressing Ctrl+Shift+J or F12.
  </p>
  <p>
    Please report this error to our support team via e-mail: 
    <a
      href="mailto:tech.support@turris.cz"
    >
      tech.support@turris.cz
    </a>
    .
  </p>
</div>
`;

exports[`<ErrorBoundary /> should render children components when there's no error 1`] = `
<div>
  <p>
    This is the content of the component.
  </p>
</div>
`;
